src/newgrf.cpp

```diff
@@ -7758,53 +7758,51 @@ static void FeatureTownName(ByteReader *buf)
 			std::string lang_name = TranslateTTDPatchCodes(grfid, lang, false, name);
 			grfmsg(6, "FeatureTownName: lang 0x%X -> '%s'", lang, lang_name.c_str());
 
-			townname->name[nb_gen] = AddGRFString(grfid, id, lang, new_scheme, false, name, STR_UNDEFINED);
+			style = AddGRFString(grfid, id, lang, new_scheme, false, name, STR_UNDEFINED);
 
 			lang = buf->ReadByte();
 		} while (lang != 0);
-		townname->id[nb_gen] = id;
-		townname->nb_gen++;
+		townname->styles.emplace_back(style, id);
 	}
 
-	byte nb = buf->ReadByte();
-	grfmsg(6, "FeatureTownName: %u parts", nb);
+	townname->partlists[id].resize(buf->ReadByte());
+	grfmsg(6, "FeatureTownName: " PRINTF_SIZE " parts", townname->partlists[id].size());
 
-	townname->nbparts[id] = nb;
-	townname->partlist[id] = CallocT<NamePartList>(nb);
+	int i = 0;
```

Comment:
i & j only used for debug messages now i think? worth giving them more friendly names?

Reply:
Yes, it's also possible to get the index from pointer arithmetic but that's a bit awful.